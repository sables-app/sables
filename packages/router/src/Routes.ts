import {
  DefaultEffectAPI,
  PayloadAction,
  SYMBOL_ROUTES_META,
} from "@sables/core";
import { capitalize } from "@sables/utils";

import { createPath } from "history";

import {
  addRoutes as addRoutesEffect,
  AddRoutesParams,
  combineHandlers,
  forwardTo,
} from "./effects.js";
import {
  assertNotWildCardRoutePath,
  assertNotWildCardRouteTemplatePath,
  assertWildCardRoutePath,
  assertWildCardRouteTemplatePath,
  createPathParser,
  ExtractParamName,
  isWildCardPath,
  normalizeRoutePath,
  PathFromParamName,
  TemplatePath,
} from "./Path.js";
import {
  cloneRouteEffects,
  createRouteEffects,
  RouteEffects,
  UntouchedRouteEffects,
} from "./RouteEffects.js";
import type {
  AnyRoutePath,
  DynamicImportFn,
  MatchingHref,
  PartialHistoryPathStrict,
  RegisterDynamicImportFn,
  RouteEffectHandlers,
  RouteHref,
  RouteID,
  RouteMiddleware,
  RouteParams,
  RoutePathType,
  RouteWithParams,
  WildCardPathType,
} from "./types.js";

/** @internal */
type ExternalEffectsFn<EffectAPI extends DefaultEffectAPI> =
  | DynamicImportFn<{ default: RouteEffects<EffectAPI> }>
  | (() => RouteEffects<EffectAPI>);

/** @internal */
type RoutesMeta<EffectAPI extends DefaultEffectAPI> = {
  externalEffectsFn?: ExternalEffectsFn<EffectAPI>;
  internalEffects: UntouchedRouteEffects<EffectAPI, any>;
  routesById: Map<string, RouteReference<RouteID>>;
};

/** @internal */
function cloneRoutesMeta<EffectAPI extends DefaultEffectAPI>(
  routesMeta: RoutesMeta<EffectAPI>
): RoutesMeta<EffectAPI> {
  return {
    externalEffectsFn: routesMeta.externalEffectsFn,
    internalEffects: cloneRouteEffects(routesMeta.internalEffects),
    routesById: new Map(routesMeta.routesById.entries()),
  };
}

/** @internal */
function cloneRoutes<
  EffectAPI extends DefaultEffectAPI,
  Info extends RouteReferenceInfo<string, string> = never
>(routes: UntouchedRoutes<EffectAPI, Info>): UntouchedRoutes<EffectAPI, Info> {
  const nextRoutes = createRoutes<EffectAPI, Info>();

  nextRoutes[SYMBOL_ROUTES_META] = cloneRoutesMeta(routes[SYMBOL_ROUTES_META]);

  const routeReferences = Object.fromEntries(
    [...nextRoutes[SYMBOL_ROUTES_META].routesById.entries()].map(
      ([id, routeReference]) => [capitalize(id), routeReference]
    )
  );

  return Object.assign(nextRoutes, routeReferences);
}

/** @internal */
export type RouteReferenceInfo<
  RID extends string,
  RawParamName extends string
> = [RID, RawParamName];

/** @internal declaration requirement */
export interface RoutesMethods<
  EffectAPI extends DefaultEffectAPI,
  Info extends RouteReferenceInfo<string, string> = never
> {
  /**
   * @public
   */
  getRouteByID<RID extends RouteID>(id?: RID): RouteReference<RID> | undefined;

  /**
   * Retrieves a map of handlers to invoke effects matching
   * the given route ID.
   *
   * @privateRemarks
   *
   * To ensure no parameters are accidentally left out,
   * a union type with `undefined` is used instead of making the parameters optional.
   *
   * @internal
   */
  _getHandlersByRouteID(
    id: RouteID | undefined,
    registerDynamicImport: RegisterDynamicImportFn
  ): Promise<RouteEffectHandlers<EffectAPI>>;

  /**
   * Retrieves a set route by matching the given hyperlink.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/")
   *   .set("profile", "/profiles/:handle");
   *
   * const profileRouteWithParams = routes.find("/profiles/@ash.ketchum");
   *
   * @public
   */
  find(href: MatchingHref): RouteWithParams | undefined;

  /**
   * Sets a route to be matched for route transitions.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/")
   *   .set("profile", "/profiles/:handle");
   *
   * @public
   */
  set<RID extends RouteID, RawParamName extends string>(
    id: RID,
    path: RoutePathType | TemplatePath<RoutePathType, RawParamName>
  ): UntouchedRoutes<EffectAPI, RouteReferenceInfo<RID, RawParamName> | Info>;

  /**
   * Sets route to forward to another location.
   *
   * @remarks
   *
   * The resulting route is considered unidentified, and a `forwardTo`
   * route effect is appended to the internal Route Effects object.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/")
   *   // Forwards "/home.html" to "/"
   *   .setForwarding("/home.html", () => "/");
   *
   * @privateRemarks
   *
   * `NavigationDestination` isn't accepted, because typings break
   * when the `Routes` object references itself.
   * If that pattern is desired, the effect should be added using
   * `forwardTo` on a Route Effects object.
   *
   * @public
   */
  setForwarding(
    path: RoutePathType,
    getDestination: (
      params: RouteParams
    ) => PartialHistoryPathStrict | RouteHref
  ): UntouchedRoutes<EffectAPI, Info>;

  /**
   * Sets a wildcard route to be matched for route transitions.
   *
   * Wildcard routes _must_ be set after non-wildcard routes.
   * TypeScript definitions enforce this pattern.
   *
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/")
   *   .setWildcard("search", "/search/*");
   *
   * @public
   */
  setWildcard<RID extends RouteID, RawParamName extends string>(
    id: RID,
    path: WildCardPathType | TemplatePath<WildCardPathType, RawParamName>
  ): Omit<
    UntouchedRoutes<EffectAPI, RouteReferenceInfo<RID, RawParamName> | Info>,
    WildcardLockedMethods
  >;

  /**
   * Sets wildcard route to add the given routes when matched.
   *
   * The given callback function should return a `Routes` object,
   * or a promise that resolves to an object with a `Routes` object
   * assigned to the `default` property. The latter can be used to lazy-load
   * routes using dynamic imports.
   *
   * @remarks
   *
   * The resulting route is considered unidentified, and an `addRoutes`
   * route effect is appended to the internal Route Effects object.
   *
   * Wildcard routes _must_ be set after non-wildcard routes.
   * TypeScript definitions enforce this pattern.
   *
   * @example
   *
   * const settingsRoutes = createRoutes()
   *   .set("notifications", "/settings/notifications")
   *   .set("resetPassword", "/settings/reset-password");
   *
   * const routes = createRoutes()
   *   .set("root", "/")
   *   .setRoutes("/settings/*", () => settingsRoutes)
   *   .setRoutes("/company/*", () => import("./companyRoutes.js"));
   *
   * @public
   */
  setRoutes(
    path: WildCardPathType,
    ...params: AddRoutesParams<EffectAPI>
  ): Omit<UntouchedRoutes<EffectAPI, Info>, WildcardLockedMethods>;

  /**
   * Sets a Route Effects object to be used during route transitions.
   * The given value may be a Route Effects object, or a callback with
   * a dynamic import to lazy-load the effects.
   *
   * @remarks
   *
   * Route Effects _must_ be set after all other method calls; the last
   * in the chain.
   * TypeScript definitions enforce this pattern.
   *
   * @example
   *
   * const routes = createRoutes()
   *   .set("root", "/")
   *   .set("profile", "/profiles/:handle")
   *   .setEffects(() => import("./routeEffects.js"));
   *
   * @public
   */
  setEffects(
    externalEffectsFn: ExternalEffectsFn<EffectAPI>
  ): Routes<EffectAPI, Info>;

  /** @internal */
  [SYMBOL_ROUTES_META]: RoutesMeta<EffectAPI>;
}

/** @internal */
export type RouteParamsFromParamName<ParamName extends string> = Record<
  ParamName,
  unknown
>;

/** @public */
export type RouteReference<
  RID extends RouteID,
  ParamName extends string = string,
  Path extends AnyRoutePath = PathFromParamName<ParamName>,
  PathParams extends RouteParamsFromParamName<ParamName> = RouteParamsFromParamName<ParamName>
> = Readonly<{
  build(params?: PathParams | null): RouteHref;
  id: RID;
  path: Path;
  test(href: MatchingHref): PathParams | null;
  match(href: MatchingHref): boolean;
  toString(): RID;
}>;

/** @public */
export type AnyRouteReference = RouteReference<RouteID>;

type RouteIDFromInfo<Info> = Info extends RouteReferenceInfo<infer RID, string>
  ? RID
  : never;
type RawParamNameFromInfo<Info> = Info extends RouteReferenceInfo<
  string,
  infer RawParamName
>
  ? RawParamName
  : never;

type RouteReferenceRecord<
  Info extends RouteReferenceInfo<string, string> = never
> = {
  readonly [K in Info as Capitalize<RouteIDFromInfo<K>>]: RouteReference<
    RouteIDFromInfo<K>,
    ExtractParamName<RawParamNameFromInfo<K>>
  >;
};

/** @internal */
type UntouchedRoutes<
  EffectAPI extends DefaultEffectAPI,
  Info extends RouteReferenceInfo<string, string> = never
> = RoutesMethods<EffectAPI, Info> & RouteReferenceRecord<Info>;

type LockedMethods = "set" | "setForwarding" | "setWildcard" | "setRoutes";
type WildcardLockedMethods = "set" | "setForwarding";

/** @public */
export type Routes<
  EffectAPI extends DefaultEffectAPI,
  Info extends RouteReferenceInfo<string, string> = never
> = Omit<UntouchedRoutes<EffectAPI, Info>, LockedMethods>;

const EMPTY_ROUTE_EFFECT_HANDLERS: RouteEffectHandlers<any> = {
  middleware: async () => undefined,
  onComplete: async () => undefined,
  onEnd: async () => undefined,
  onExit: async () => undefined,
  onFailure: async () => undefined,
  onInterrupt: async () => undefined,
  onStart: async () => undefined,
};

const WILDCARD_ROUTES_ID_PREFIX = "__wildcardRoutes:";

/**
 * Creates a map of routes to enable route transitions.
 *
 * @example
 *
 * const routes = createRoutes()
 *   .set("root", "/")
 *   .set("about", "/about-us")
 *   // Lazy-load additional routes
 *   .setRoutes("/users/*", () => import("./routes/users.js"));
 *
 * @public
 */
export function createRoutes<
  EffectAPI extends DefaultEffectAPI,
  Info extends RouteReferenceInfo<string, string> = never
>(): UntouchedRoutes<EffectAPI, Info> {
  const routes = {
    set(id, pathInput) {
      const templatePath = normalizeRoutePath(pathInput);

      assertNotWildCardRouteTemplatePath(templatePath);

      return addRoute(id, templatePath);
    },
    setForwarding(path, getDestination) {
      assertNotWildCardRoutePath(path);

      return addRouteWithEffect(path, forwardTo<EffectAPI>(getDestination));
    },
    setWildcard(id, pathInput) {
      const templatePath = normalizeRoutePath(pathInput);

      assertWildCardRouteTemplatePath(templatePath);

      return addRoute(id, templatePath);
    },
    setRoutes(path, ...addRoutesParams) {
      assertWildCardRoutePath(path);

      return addRouteWithEffect(
        path,
        addRoutesEffect<EffectAPI>(...addRoutesParams)
      );
    },
    setEffects(externalEffectsFn) {
      const nextRoutes = cloneRoutes(routes);

      nextRoutes[SYMBOL_ROUTES_META].externalEffectsFn = externalEffectsFn;

      return nextRoutes;
    },
    find(href: MatchingHref) {
      if (!href) return undefined;

      const routesList = routes[SYMBOL_ROUTES_META].routesById.values();

      for (const route of routesList) {
        const { id, path } = route;
        const params = route.test(href);

        if (params !== null) {
          return { id, params, path };
        }
      }
    },
    async _getHandlersByRouteID(id, registerDynamicImport) {
      if (!id) {
        return EMPTY_ROUTE_EFFECT_HANDLERS;
      }

      const { internalEffects, externalEffectsFn } = routes[SYMBOL_ROUTES_META];
      const internalHandlers = await internalEffects._getHandlersByRouteID(id);
      const externalEffects = await resolveExternalEffects(
        externalEffectsFn,
        registerDynamicImport
      );
      const externalHandlers = await externalEffects?._getHandlersByRouteID(id);

      if (!externalHandlers) {
        return internalHandlers;
      }

      return combineHandlers(internalHandlers, externalHandlers);
    },
    getRouteByID(id) {
      if (!id) return;

      return routes[SYMBOL_ROUTES_META].routesById.get(id);
    },
    [SYMBOL_ROUTES_META]: {
      routesById: new Map(),
      internalEffects: createRouteEffects<EffectAPI>(),
    },
  } as UntouchedRoutes<EffectAPI, Info>;

  function addRoute<
    RID extends RouteID,
    Path extends AnyRoutePath,
    ParamName extends string
  >(id: RID, templatePath: TemplatePath<Path, ParamName>) {
    const referenceKey = capitalize(id);
    const parser = createPathParser(templatePath);

    type PathParams = RouteParamsFromParamName<ParamName>;

    function build(params?: PathParams | null) {
      return parser.build(params || {}) as RouteHref;
    }

    function resolveHref(value: NonNullable<MatchingHref>): string {
      return typeof value === "string" ? value : createPath(value);
    }

    function test(href: MatchingHref): PathParams | null {
      if (!href) return null;

      const [hrefWithoutHash] = resolveHref(href).split("#");

      return parser.test(hrefWithoutHash) as PathParams | null;
    }

    function match(href: MatchingHref): boolean {
      return !!test(href);
    }

    const routeReference: RouteReference<RID> = {
      build,
      id,
      match,
      path: templatePath.path,
      test,
      toString: () => id,
    };

    const nextRoutes = cloneRoutes(routes);

    Object.assign(nextRoutes, { [referenceKey]: routeReference });
    nextRoutes[SYMBOL_ROUTES_META].routesById.set(id, routeReference);

    return nextRoutes;
  }

  function addRouteWithEffect<
    MW extends RouteMiddleware<PayloadAction<any>, EffectAPI>
  >(
    path: RoutePathType | WildCardPathType,
    middleware: MW
  ): UntouchedRoutes<EffectAPI, Info>;
  function addRouteWithEffect<
    MW extends RouteMiddleware<PayloadAction<any>, EffectAPI>
  >(
    path: WildCardPathType,
    middleware: MW
  ): Omit<UntouchedRoutes<EffectAPI, Info>, WildcardLockedMethods>;
  function addRouteWithEffect<
    MW extends RouteMiddleware<PayloadAction<any>, EffectAPI>
  >(path: RoutePathType | WildCardPathType, middleware: MW) {
    // Routes are typically created upon execution,
    // so this ID must be a deterministic to be used
    // in Cloudflare Workers.
    const routeID = `${WILDCARD_ROUTES_ID_PREFIX}${path}`;

    const nextRoutes:
      | UntouchedRoutes<EffectAPI, Info>
      | Omit<UntouchedRoutes<EffectAPI, Info>, WildcardLockedMethods> =
      isWildCardPath(path)
        ? routes.setWildcard(routeID, path)
        : routes.set(routeID, path);

    const nextRouteMeta = nextRoutes[SYMBOL_ROUTES_META];
    const nextInternalEffects = nextRouteMeta.internalEffects;

    nextRouteMeta.internalEffects = nextInternalEffects.append(
      routeID,
      middleware
    );

    return nextRoutes;
  }

  async function resolveExternalEffects(
    externalEffectsFn: ExternalEffectsFn<EffectAPI> | undefined,
    registerDynamicImport: RegisterDynamicImportFn
  ): Promise<RouteEffects<EffectAPI> | undefined> {
    if (!externalEffectsFn) return undefined;

    const result = externalEffectsFn();
    const routeEffects =
      result instanceof Promise ? (await result).default : result;

    function isDynamicImportFn(
      _value: typeof externalEffectsFn
    ): _value is DynamicImportFn<{ default: RouteEffects<EffectAPI> }> {
      return result instanceof Promise;
    }

    if (isDynamicImportFn(externalEffectsFn)) {
      registerDynamicImport(externalEffectsFn);
    }

    return routeEffects;
  }

  return routes;
}

/**
 * Determines whether the given object is a Routes instance.
 *
 * @see {createRoutes}
 *
 * @public
 */
export function isRoutes<
  EffectAPI extends DefaultEffectAPI,
  Info extends RouteReferenceInfo<string, string>
>(value: unknown): value is Routes<EffectAPI, Info> {
  return (
    typeof value == "object" &&
    value !== null &&
    Object.hasOwn(value, SYMBOL_ROUTES_META)
  );
}
