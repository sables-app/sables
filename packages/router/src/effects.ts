import { AnyPayloadAction, DefaultEffectAPI, SideEffect } from "@sables/core";
import { SYMBOL_EFFECT_API_EFFECT_STATE } from "@sables/core";
import { isSSREnv } from "@sables/utils";

import type { Routes } from "./Routes.js";
import type {
  BuildHrefInput,
  DynamicImportFn,
  NavigationDestination,
  RouteEffectHandlers,
  RouteListener,
  RouteMiddleware,
  RouteParams,
  StartTransitionAction,
} from "./types.js";
import {
  createDynamicImportRegistrar,
  demandRoutesCollectionFromEffectAPI,
} from "./utils.js";

function getEffectStateMap<
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
>(effectAPI: EffectAPI, id: symbol) {
  if (!effectAPI[SYMBOL_EFFECT_API_EFFECT_STATE]) {
    effectAPI[SYMBOL_EFFECT_API_EFFECT_STATE] = new Map();
  }

  return effectAPI[SYMBOL_EFFECT_API_EFFECT_STATE];
}

const effectState = {
  has<EffectAPI extends DefaultEffectAPI = DefaultEffectAPI>(
    effectAPI: EffectAPI,
    id: symbol
  ) {
    return getEffectStateMap(effectAPI, id).has(id);
  },
  get<EffectAPI extends DefaultEffectAPI = DefaultEffectAPI>(
    effectAPI: EffectAPI,
    id: symbol
  ) {
    return getEffectStateMap(effectAPI, id).get(id);
  },
  set<EffectAPI extends DefaultEffectAPI = DefaultEffectAPI>(
    effectAPI: EffectAPI,
    id: symbol,
    value: unknown
  ) {
    getEffectStateMap(effectAPI, id).set(id, value);
  },
};

function createEffectState<T>(defaultValue?: T) {
  const effectID = Symbol();

  return {
    using: <EffectAPI extends DefaultEffectAPI = DefaultEffectAPI>(
      effectAPI: EffectAPI
    ) => {
      function get() {
        if (!effectState.has(effectAPI, effectID)) {
          return defaultValue;
        }
        return effectState.get(effectAPI, effectID) as T | undefined;
      }
      function set(value: T) {
        effectState.set(effectAPI, effectID, value);
      }
      return Object.assign([get(), set] as const, { get, set });
    },
  };
}

/**
 * The base class for all route middleware signals.
 *
 * A route middleware signal is an object thrown within a
 * route middleware function to halt a route transition.
 *
 * @remarks
 *
 * `RouteMiddlewareSignal` doesn't cause any effects by default.
 * It should only be used for type checking.
 *
 * @example
 *
 * value instanceof RouteMiddlewareSignal;
 *
 * @public
 */
export class RouteMiddlewareSignal {
  constructor(public message = "A route middleware function was halted.") {}
}

/**
 * A route middleware signal thrown to add routes to the application.
 *
 * @remarks
 *
 * It's recommended to use the `addRoutes` middleware creator when applicable.
 *
 * @see {RouteMiddlewareSignal}
 * @see {addRoutes}
 *
 * @public
 */
export class AddRoutesSignal<
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
> extends RouteMiddlewareSignal {
  constructor(
    public routes: Routes<EffectAPI>,
    message = "Attempting to add routes."
  ) {
    super(message);
  }
}

/**
 * A route middleware signal thrown to forward a route transition to another location.
 *
 * @remarks
 *
 * It's recommended to use the `forwardTo` route middleware creator when applicable.
 *
 * @see {RouteMiddlewareSignal}
 * @see {forwardTo}
 *
 * @public
 */
export class ForwardRouteSignal<
  Route extends BuildHrefInput
> extends RouteMiddlewareSignal {
  constructor(
    public destination: NavigationDestination<Route>,
    message = "Forwarding route."
  ) {
    super(message);
  }
}

/**
 * A route middleware signal thrown to preemptively end a route transition.
 *
 * @remarks
 *
 * It's recommended to use the `exitTransition` route middleware when applicable.
 *
 * @see {RouteMiddlewareSignal}
 * @see {exitTransition}
 *
 * @public
 */
export class ExitTransitionSignal extends RouteMiddlewareSignal {
  constructor(message = "Ending the route transition.") {
    super(message);
  }
}

/**
 * Route middleware that preemptively ends a route transition.
 * The middleware resolves early if the route transition was aborted.
 *
 * @example
 *
 * const routeEffects = createRouteEffects()
 *   // Prevent all route transitions
 *   .prependAll(exitTransition);
 *
 * @public
 */
export async function exitTransition<
  Action extends StartTransitionAction = StartTransitionAction,
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
>(_action: Action, _effectAPI: EffectAPI, abortSignal: AbortSignal) {
  if (abortSignal.aborted) return;
  // This is caught by `routeTransitionMiddleware` to exit the transition
  throw new ExitTransitionSignal();
}

/**
 * Creates middleware that delays a route transition for a duration.
 * The middleware resolves early if the route transition was aborted.
 * The middleware immediately resolves during SSR.
 *
 * @param ms Duration in milliseconds
 *
 * @example
 *
 * const routeEffects = createRouteEffects()
 *   // Delay all routes by five seconds.
 *   .appendAll(delayTransition(5000));
 *
 * @public
 */
export function delayTransition<
  Action extends StartTransitionAction = StartTransitionAction,
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
>(ms: number): RouteMiddleware<Action, EffectAPI> {
  return async (_action, _effectAPI, abortSignal) => {
    if (isSSREnv() || abortSignal.aborted) return;

    await Promise.race([
      new Promise((resolve) => abortSignal.addEventListener("abort", resolve)),
      new Promise((resolve) => setTimeout(resolve, ms)),
    ]);
  };
}

/**
 * Creates middleware that logs values.
 * The middleware resolves early if the route transition was aborted.
 *
 * @example
 *
 * const routeEffects = createRouteEffects()
 *   // Log a message during every route transition
 *   .appendAll(logTransition("A route transition is occurring."));
 *
 * @public
 */
export function logTransition<
  Action extends StartTransitionAction = StartTransitionAction,
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
>(...values: unknown[]): RouteMiddleware<Action, EffectAPI> {
  return async (_action, _effectAPI, abortSignal) => {
    if (abortSignal.aborted) return;
    console.log(...values);
  };
}

/**
 * Creates middleware that forwards a route transition to another location.
 * The middleware resolves early if the route transition was aborted.
 *
 * @example
 *
 * const routes = createRoutes()
 *   .set("root", "/")
 *   .set("appRoot", "/app");
 *
 * const routeEffects = createRouteEffects()
 *   // Forwards "/" to "/app"
 *   .append(routes.Root.id, forwardTo(() => routes.AppRoot));
 *
 * @public
 */
export function forwardTo<
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI,
  Route extends BuildHrefInput = BuildHrefInput
>(
  getDestination: (params: RouteParams) => NavigationDestination<Route>
): RouteMiddleware<StartTransitionAction, EffectAPI> {
  return async (action, effectAPI, abortSignal) => {
    if (abortSignal.aborted) return;

    const routesCollection = demandRoutesCollectionFromEffectAPI(effectAPI);
    const { nextRoute } = routesCollection.findByHref(
      action.payload.locationChange.location
    );
    const nextRouteParams = nextRoute?.params || {};
    const destination = getDestination(nextRouteParams);

    // This is caught by `routeTransitionMiddleware` to redirect the transition
    throw new ForwardRouteSignal(destination);
  };
}

/**
 * @public
 */
export type AddRoutesParams<EffectAPI extends DefaultEffectAPI> = [
  routes:
    | (() => Routes<EffectAPI>)
    | DynamicImportFn<{ default: Routes<EffectAPI> }>,
  message?: string
];

/**
 * Creates middleware that adds the given routes to the application.
 * The middleware resolves early if the route transition was aborted.
 *
 * @remarks
 *
 * Routes can only be added during a route transition.
 * Whenever routes are added, the current route transition is interrupted,
 * and a new route transition is started.
 *
 * @example
 *
 * const initialRoutes = createRoutes()
 *   .set("root", "/")
 *   .setWildcard("profile", "/profile/*");
 *
 * const profileRoutes = createRoutes()
 *   .set("profile", "/profile/:handle")
 *   .set("profileSettings", "/profile/settings");
 *
 * const routeEffects = createRouteEffects()
 *   // Add `profileRoutes` when the location matches `initialRoutes.Profile`
 *   .append(
 *     initialRoutes.Profile.id,
 *     addRoutes(() => profileRoutes)
 *   );
 *
 * initialRoutes.setEffects(() => routeEffects);
 *
 * @public
 */
export function addRoutes<
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
>(
  ...params: AddRoutesParams<EffectAPI>
): RouteMiddleware<StartTransitionAction, EffectAPI> {
  const [routesFn, message] = params;
  const routesWereAddedState = createEffectState(false);

  return async function addRoutesEffect(...params) {
    const [, effectAPI, abortSignal] = params;
    const [routesWereAdded, setRoutesWereAdded] =
      routesWereAddedState.using(effectAPI);

    if (routesWereAdded) return;

    setRoutesWereAdded(true);

    if (abortSignal.aborted) return;

    const result = routesFn();
    const routes = result instanceof Promise ? (await result).default : result;

    function isDynamicImportFn(
      _value: typeof routesFn
    ): _value is DynamicImportFn<{ default: Routes<EffectAPI> }> {
      return result instanceof Promise;
    }

    if (isDynamicImportFn(routesFn)) {
      createDynamicImportRegistrar(effectAPI)(routesFn);
    }

    throw new AddRoutesSignal(routes, message);
  };
}

/**
 * Creates a new route middleware function, that calls the given
 * middleware in an asynchronous series.
 * The series execution is halted if the transition is aborted.
 *
 * @example
 *
 * const myRouteEffect = chainMiddleware(
 *   delayTransition(3000),
 *   logTransition("A route transition occurred."),
 * );
 *
 * @public
 */
export function chainMiddleware<
  Action extends StartTransitionAction = StartTransitionAction,
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
>(
  ...effects: RouteMiddleware<Action, EffectAPI>[]
): RouteMiddleware<Action, EffectAPI> {
  return async (action, effectAPI, abortSignal) => {
    for (const effect of effects) {
      if (abortSignal.aborted) return;

      await effect(action, effectAPI, abortSignal);
    }
  };
}

/**
 * Creates a new route listener that calls the given
 * route listeners in series synchronously.
 *
 * @example
 *
 * const myRouteListener = combineListeners(
 *   () => console.log("First log."),
 *   () => console.log("Second log."),
 * );
 *
 * @public
 */
export function combineListeners<
  Action extends AnyPayloadAction = AnyPayloadAction,
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
>(
  ...listeners: RouteListener<Action, EffectAPI>[]
): RouteListener<Action, EffectAPI> {
  return (action, effectAPI) => {
    for (const listener of listeners) {
      listener(action, effectAPI);
    }
  };
}

/** @internal */
export function combineHandlers<EffectAPI extends DefaultEffectAPI>(
  ...handlersCollection: RouteEffectHandlers<EffectAPI>[]
): RouteEffectHandlers<EffectAPI> {
  // prettier-ignore
  return {
    middleware:  chainMiddleware(...handlersCollection.map(({ middleware })   => middleware)),
    onComplete:  combineListeners(...handlersCollection.map(({ onComplete })  => onComplete)),
    onEnd:       combineListeners(...handlersCollection.map(({ onEnd })       => onEnd)),
    onExit:      combineListeners(...handlersCollection.map(({ onExit })      => onExit)),
    onFailure:   combineListeners(...handlersCollection.map(({ onFailure })   => onFailure)),
    onInterrupt: combineListeners(...handlersCollection.map(({ onInterrupt }) => onInterrupt)),
    onStart:     combineListeners(...handlersCollection.map(({ onStart })     => onStart)),
  };
}

/**
 * Creates a new route middleware using the given side effect.
 *
 * @example
 *
 * const getBooks = createSideEffect(
 *   createAction<{ id: string }>("getBooks/start"),
 *   createAction<Response>("getBooks/end"),
 *   async (action) =>
 *     fetch(`https://example.com/books/${action.payload.id}`)
 * );
 * const loadBooks = sideEffectToRouteMiddleware(getBooks, (params) =>
 *   yup.object({ id: yup.string().required() }).validate(params)
 * );
 *
 * // Loads books for every route transition
 * createRoutes().setEffects(() => createRouteEffects().appendAll(loadBooks));
 *
 * @public
 */
export function sideEffectToRouteMiddleware<
  EffectAPI extends DefaultEffectAPI,
  S extends SideEffect<any, any, EffectAPI>
>(
  sideEffect: S,
  assertRouteParams?: (
    params: RouteParams
  ) => asserts params is ReturnType<S["actions"]["start"]>["payload"]
): RouteMiddleware<StartTransitionAction, EffectAPI> {
  return async (action, effectAPI, abortSignal) => {
    if (abortSignal.aborted) return;

    const params = action.payload.nextRoute?.params || {};

    await assertRouteParams?.(params);

    return sideEffect(effectAPI, params);
  };
}
