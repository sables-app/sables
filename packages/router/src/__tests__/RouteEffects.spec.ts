import { assertType, describe, expect, it, test } from "vitest";

import { delayTransition } from "../effects.js";
import { createRouteEffects } from "../RouteEffects.js";
import { createRoutes } from "../Routes.js";

describe("RoutesEffects", () => {
  describe("createRoutesEffects", () => {
    test.only("appending route middleware", () => {
      const routes = createRoutes()
        .set("root", "/")
        .set("profile", "/profiles/:handle")
        .set("what", "/what");

      const routeEffects = createRouteEffects()
        // Append a route effect using a route reference
        .append(routes.Root, delayTransition(0))
        // Append a route effect using a route ID
        .append(routes.Profile.id, delayTransition(0))
        // Append a second route effect to the same route using a route reference
        .append(routes.Profile, delayTransition(0))
        // Add a transition complete listener using a route reference
        .onComplete(routes.Profile, () => undefined);

      expect(routeEffects).toMatchSnapshot();

      assertType<"root">(routeEffects.handlers.RootMiddleware.id);
      expect(routeEffects.handlers.RootMiddleware.id).toEqual("root");

      assertType<"profile">(routeEffects.handlers.ProfileMiddleware.id);
      expect(routeEffects.handlers.ProfileMiddleware.id).toEqual("profile");
    });
  });
});
