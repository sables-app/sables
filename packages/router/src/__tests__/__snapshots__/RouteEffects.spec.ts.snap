// Vitest Snapshot v1

exports[`RoutesEffects > createRoutesEffects > appending route middleware 1`] = `
{
  "_clone": [Function],
  "_getHandlersByRouteID": [Function],
  "append": [Function],
  "appendAll": [Function],
  "handlers": {
    "ProfileMiddleware": [Function],
    "RootMiddleware": [Function],
  },
  "onComplete": [Function],
  "onCompleteAll": [Function],
  "onEnd": [Function],
  "onEndAll": [Function],
  "onExit": [Function],
  "onExitAll": [Function],
  "onFailure": [Function],
  "onFailureAll": [Function],
  "onInterrupt": [Function],
  "onInterruptAll": [Function],
  "onStart": [Function],
  "onStartAll": [Function],
  "prepend": [Function],
  "prependAll": [Function],
  Symbol(@sables/routeEffectsMeta): {
    "listeners": Set {
      {
        "listenerReference": [Function],
      },
    },
    "middleware": Set {
      {
        "middlewareReference": [Function],
        "referenceKey": "RootMiddleware",
      },
      {
        "middlewareReference": [Function],
        "referenceKey": "ProfileMiddleware",
      },
      {
        "middlewareReference": [Function],
        "referenceKey": "ProfileMiddleware",
      },
    },
  },
}
`;
